candle-nn/src/encoding.rs

```diff
@@ -0,0 +1,379 @@
+///
+/// let expected_matrix = [
+///     [[1.0, 0.0, 0.0, 0.0], [0.0, 0.0, 1.0, 0.0]],
+///     [[0.0, 1.0, 0.0, 0.0], [0.0, 0.0, 0.0, 0.0]],
+/// ];
+///
+/// assert_eq!(one_hot.shape(), &Shape::from((2, 2, depth)));
+///
+/// let matrix = one_hot.to_vec3::<f32>()?;
+///
+/// assert_eq!(matrix, expected_matrix);
+///```
+/// ## One-cold Encoding
+///
+/// ```rust
+/// use candle::{Shape, Tensor, Device};
+/// use candle_nn::encoding::one_hot;
+///
+///
+/// let device = candle::Device::Cpu;
+/// let depth = 4;
+/// let indices = Tensor::new(vec![vec![0i64, 2], vec![1, 3]], &device)?;
+///
+/// let allow_f64 = false;
+/// let on_value = Some(0);
+/// let off_value = Some(1);
+///
+/// let one_cold = one_hot::<i64, u8>(indices, depth, allow_f64, on_value, off_value)?;
+///
+/// let expected_matrix = [[[0, 1, 1, 1], [1, 1, 0, 1]], [[1, 0, 1, 1], [1, 1, 1, 0]]];
+///
+/// assert_eq!(one_cold.shape(), &Shape::from((2, 2, depth)));
+///
+/// let matrix = one_cold.to_vec3::<u8>()?;
+///
+/// assert_eq!(matrix, expected_matrix);
+/// ```
+///
+///
+/// # Bails
+///
+/// This method will bail on tensors with a rank greater than 3.
+///
+/// This method will bail if the index value is less than -1.
+///
+/// This method will bail if the index value is greater than or equal to the depth value.
+///
+/// This method will bail if the input data type does not equal the generic data type provided.
+///
+/// This method will bail if the input data type is `f64` and `allow_f64` is false.
+///
+/// # API Design
+///
+/// The api design for this method is loosely based on the [TensorFlow One-Hot](https://www.tensorflow.org/api_docs/python/tf/one_hot) method.
+pub fn one_hot<I: WithDType, O: WithDType>(
+    indices: Tensor,
+    depth: usize,
+    allow_f64: bool,
+    on_value: Option<O>,
+    off_value: Option<O>,
+) -> Result<Tensor> {
+    let on_value = on_value.unwrap_or(O::from_f64(1.));
+    let off_value = off_value.unwrap_or(O::from_f64(0.));
+
+    if I::DTYPE != indices.dtype() {
+        bail!(
+            "one_hot: indices dtype {} does not match expected dtype {}",
+            indices.dtype().as_str(),
+            I::DTYPE.as_str()
+        )
+    }
+
+    if !allow_f64 && I::DTYPE.as_str() == "f64" {
+        bail!("one_hot: f64 is not supported")
+    }
+
+    let rank = indices.rank();
+    match rank {
+        0 => {
+            let mut v = vec![off_value; depth];
+            let index = indices.to_vec0::<I>()?;
+            let vi = index.to_f64();
+
+            check_value(vi, depth)?;
+
+            if vi >= 0. {
+                v[vi as usize] = on_value;
+            }
+
+            Tensor::new(v, indices.device())
+        }
+        1 => {
+            let dim1 = indices.dims1()?;
+            let mut v = vec![off_value; depth * dim1];
+            let index = indices.to_vec1::<I>()?;
+            for i in 0..dim1 {
+                let vi = index[i].to_f64();
```

Comment:
Same comment as earlier, all this code should be tweaked to manipulate integers rather than floats for indexes so as to avoid numerical precision issues (even if they are impossible to happen for small indexes values).